### PR TITLE
Handle the proto-copying with Go

### PR DESCRIPTION
**Pull request by jrockway**
People are having trouble generating the protos on Mac because `cp` behaves differently compared to Linux.  This is annoying, so I have moved most of the logic to Go; we still have a shell script to run protoc and gopatch, but we do everything else with a custom program.  We have also had trouble with GNU tar vs. BSD tar before, so again, I avoided trying to invoke tar in the shell script and just use archive/tar.

The actual proto generation now runs as a genrule, which means that it's cached when the protos don't change.

I added a "test" mode and a test to check that your working copy matches what came out of the genrule.  If it doesn't match, then run "extract" (still aliased to //:make_proto).  The test output, and the output of make proto, look something like this:

```
INFO prototar/main.go:187 updated file {"name": "proto-docs.json"}
INFO prototar/main.go:187 updated file {"name": "proto-docs.md"}
INFO prototar/main.go:187 updated file {"name": "src/internal/jsonschema/admin_v2/ClusterInfo.schema.json"}
INFO prototar/main.go:187 updated file {"name": "src/openapi/pachyderm_api.swagger.json"}
INFO prototar/main.go:187 updated file {"name": "src/typescript/admin/admin.pb.ts"}
INFO prototar/main.go:187 updated file {"name": "src/admin/admin.pb.go"}
INFO prototar/main.go:187 updated file {"name": "src/admin/admin.pb.validate.go"}
INFO prototar/main.go:187 updated file {"name": "src/admin/admin.pb.zap.go"}

*** It looks like you might need to regenerate the protos in your working copy.
```

It will also print files that are unchanged (in debug mode) and files that it's deleting (when you remove a message from a proto and the JSON schema file needs to be killed).

I added a feature that lets you see "forgotten" files that you might have forgotten to copy over because they were generated in a weird location.  Do something like `bazel build //src/proto:gen_go_proto_bundle; cat bazel-bin/src/proto/go_protos_forgotten_files.txt` to see them.  This looks something like:

```
out/github.com/chrusty/protoc-gen-jsonschema/json-schema-options.pb.go
out/github.com/chrusty/protoc-gen-jsonschema/json-schema-options.pb.validate.go
out/github.com/chrusty/protoc-gen-jsonschema/json-schema-options.pb.zap.go
out/github.com/envoyproxy/protoc-gen-validate/validate/validate.pb.go
out/github.com/envoyproxy/protoc-gen-validate/validate/validate.pb.validate.go
out/github.com/envoyproxy/protoc-gen-validate/validate/validate.pb.zap.go
```

Indeed, we don't want those.  But now we can be sure that nothing is being forgotten, which will be helpful when someone adds a weird plugin.

### CODE REVIEW

File: src/internal/log/loggers.go [hunk 1]
```diff
@@ -214,12 +214,12 @@ func InitBatchLogger(logFile string) func(err error) {
 			if logFile == "" {
 				var err error
 				// On error, out is set to "" again.
-				logFile, err = xdg.CacheFile(fmt.Sprintf("pachyderm/log/%s.%s.log", name, time.Now().In(time.UTC).Format("20060102T150405Z")))
+				name := fmt.Sprintf("%s.%s.log", name, time.Now().In(time.UTC).Format("20060102T150405Z"))
```
Comment (jrockway) [inline src/internal/log/loggers.go:217]: Not allowed to write to ~/.cache in a genrule.

File: src/internal/pctx/context.go [hunk 2]
```diff
@@ -21,6 +23,11 @@ func Background(process string) context.Context {
 	return Child(ctx, process)
 }
 
+// Interactive returns a context for use in command-line apps.
+func Interactive() (context.Context, context.CancelFunc) {
+	return signal.NotifyContext(log.AddLogger(context.Background()), os.Interrupt)
```
Comment (jrockway) [inline src/internal/pctx/context.go:28]: I was tired of copy-pasting this.  Also terminationSignals is very annoying to use, because it ignores C-\, so there is no way to shut down a broken program.  For that reason, we only grab interrupt here.

File: src/proto/test.sh [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+# shellcheck source=/dev/null.
+source "${RUNFILES_DIR:-/dev/null}/$f" 2>/dev/null || \
+source "$(grep -sm1 "^$f " "${RUNFILES_MANIFEST_FILE:-/dev/null}" | cut -f2- -d' ')" 2>/dev/null || \
+source "$0.runfiles/$f" 2>/dev/null || \
+source "$(grep -sm1 "^$f " "$0.runfiles_manifest" | cut -f2- -d' ')" 2>/dev/null || \
+source "$(grep -sm1 "^$f " "$0.exe.runfiles_manifest" | cut -f2- -d' ')" 2>/dev/null || \
+{ echo>&2 "ERROR: cannot find $f; this script must be run with 'bazel run'"; exit 1; }; f=; set -e
+# --- end runfiles.bash initialization v3 ---
+
+cd ../../../../../../..
```
Comment (jrockway) [inline src/proto/test.sh:18]:
```
¯\_(ツ)_/¯
```

File: src/internal/log/loggers.go (hunk 1 above)
Comment (robert-uhl) [reply src/internal/log/loggers.go:217]: Could you set `XDG_CACHE_HOME` when this is called from a genrule?

File: src/internal/pctx/context.go (hunk 2 above)
Comment (robert-uhl) [reply src/internal/pctx/context.go:28]: It’s great to fix this instead of copy-pasting.

I guess it’s _okay_ not to take a context as an argument, since signals are process-global, and the user can always augment the context later — unless using another library which thinks its context should be global for another reason. Might still be better to take an argument.

What signal does `C-,` send in your terminal? `signals.TerminationSignals` is (on Unix) `SIGINT`, `SIGTERM`, `SIGQUIT` and `SIGHUP`. I guess your terminal probably sends `SIGQUIT`, which is normally bound to `C-\`? One can always shut down a broken program with `kill -9`, of course, but I see your point.

Maybe this should be smarter still:

- `SIGINT`/`SIGHUP` (normally on `C-c`): cancel the context
- `SIGQUIT`: `os.Exit(255)`

File: src/proto/extract.sh [hunk 4]
```diff
@@ -0,0 +1,20 @@
+#!/usr/bin/env bash
+
+set -euo pipefail
+IFS=$'\n\t'
+
+# --- begin runfiles.bash initialization v3 ---
+# Copy-pasted from the Bazel Bash runfiles library v3.
```
Comment (robert-uhl) [inline src/proto/extract.sh:7]: The Bazel Bash runfiles library v3 is subject to [the Apache License v2](https://www.apache.org/licenses/LICENSE-2.0), which requires, among other things, a copy of the license and retention of copyright notices. I see that there is a copyright notice at the top of https://github.com/bazelbuild/bazel/blob/master/tools/bash/runfiles/runfiles.bash#L1

Why does this have to be copy-pasted rather than used?

File: src/proto/prototar/main.go [hunk 5]
```diff
@@ -0,0 +1,351 @@
...
+	}
+	if w.names == nil {
+		w.names = make(map[string]struct{})
+	}
+	w.names[path.Join(w.prefix, h.Name)] = struct{}{}
+	return nil
+}
+
+func addToArchive(ctx context.Context, tw *nameAndTarWriter, f fs.FS) error {
+	// TODO(jrockway): For this to be hermetic, WalkDir needs to always visit files in the same
```
Comment (robert-uhl) [inline src/proto/prototar/main.go:41]: Simple enough to build up the files in memory and sort them, right?

File: src/proto/test.sh [hunk 6]
```diff
@@ -0,0 +1,19 @@
+#!/usr/bin/env bash
+
+set -euo pipefail
+IFS=$'\n\t'
+
+# --- begin runfiles.bash initialization v3 ---
```
Comment (robert-uhl) [inline src/proto/test.sh:6]: Missing copyright notice here too.

**Review by robert-uhl — COMMENTED** [review]
> People are having trouble generating the protos on Mac because `cp` behaves differently compared to Linux. This is annoying, so I have moved most of the logic to Go; we still have a shell script to run `protoc` and `gopatch`, but we do everything else with a custom program. We have also had trouble with GNU `tar` vs. BSD `tar` before, so again, I avoided trying to invoke `tar` in the shell script and just use `archive/tar`.

Can we get the behavior we want from POSIX [`cp`](https://pubs.opengroup.org/onlinepubs/9699919799/utilities/cp.html) and [`tar`](https://pubs.opengroup.org/onlinepubs/7908799/xcu/tar.html) (or even [`pax`](https://pubs.opengroup.org/onlinepubs/9699919799/utilities/pax.html))? Both macOS’s BSD userland and Linux’s GNU userland should be mostly POSIX-compatible.

As a general rule, I believe that it is preferable to use POSIX rather than re-roll this stuff at the Go level. Granted, POSIX is definitely kinda crufty, and the Go standard libraries can be pretty awesome, but even in 2024 I feel that the POSIX utilities have had more thorough use and testing over the years. Shell _is_ a pretty awful programming language, but it is typically more concise. And to a first approximation bugs per SLOC are more-or-less constant, so writing hundreds of lines of Go to replace a handful of lines of shell is likely to result in increased maintenance burden long-term.

Bottom line: if we _must_ to push the logic into Go, okay, but … must we?

**Comment by smalyala** [comment]
Works as expected on Mac 👍

File: src/proto/extract.sh (hunk 4 above)
Comment (acohen4) [reply src/proto/extract.sh:7]: We can proceed with this PR, as I look over whether we could handle this differently 👍

**Review by acohen4 — APPROVED** [review]
LGTM

File: src/internal/log/loggers.go (hunk 1 above)
Comment (jrockway) [reply src/internal/log/loggers.go:217]: It seems like we can't.  Overall, I like having the log when "all else fails" even if you have to delete spurious files ("git clean" saves you).

File: src/internal/pctx/context.go (hunk 2 above)
Comment (jrockway) [reply src/internal/pctx/context.go:28]: I don't like messing with SIGQUIT.  SIGQUIT is how you kill stuff with your keyboard when all else fails.

File: src/proto/prototar/main.go (hunk 5 above)
Comment (jrockway) [reply src/proto/prototar/main.go]: Done.
